### PR TITLE
Stats

### PR DESCRIPTION
**Pull request by nick-Ag**
Implementation of running averages for each dataset channel. Most of the work done here was backend improvements and additions. The dataset component will send a POST request to the backend containing the data for which statistics should be computed (each data channel). The server will respond with the computed data for the frontend to plot.
Not Included in this PR for next PRs:

- Toggling of stats traces.

- Standard Deviation plots.

### CODE REVIEW

File: backend/server.py [hunk 1]
```diff
@@ -40,7 +43,31 @@ def upload_file():
         
         samples = GoogleSensorParser([f.filename]).parse_files()
 
-        return samples
+        return {'type': 'upload', 'data': samples}
+
+@app.route('/stats', methods = ['POST'])
+def compute_stats():
```
Comment (rglasser) [inline backend/server.py:49]: Document your functions

File: backend/server.py [hunk 2]
```diff
@@ -40,7 +51,49 @@ def upload_file():
...
+"""Computes the running average of a single data trace.
+    Returns a Python list.
+
+Attributes:
+    trace: The Python list data trace to compute averages for.
+"""
+def compute_running_avg(trace):
+    n = len(trace)
+    avgs = np.cumsum(trace[:n+1], dtype=float)
+    for i in range(len(avgs)):
```
Comment (agarwal-v) [inline backend/server.py:85]: This is cumulative average, not running average

File: frontend/src/app/dataset/dataset.component.html [hunk 3]
```diff
@@ -22,22 +22,32 @@
...
+            <mat-button-toggle (click)="showOptions('latencies')"
             *ngIf="hasLatencies">
                 Latency
             </mat-button-toggle>
-            <mat-button-toggle (click)="toggleTrace('ts_diff')">
+            <mat-button-toggle (click)="showOptions('ts_diff')">
                 Timestamp Difference
             </mat-button-toggle>
         </mat-button-toggle-group>
+        <!-- This mat-expansion-panel is reused for every data trace contained in the dataset. To maintain a memory of which
```
Comment (agarwal-v) [inline frontend/src/app/dataset/dataset.component.html:39]: What is the max line length used in the code? This is 100+ chars

File: frontend/src/app/side-menu/side-menu.component.ts [hunk 4]
```diff
@@ -59,7 +59,47 @@ export class SideMenuComponent {
...
+
+              this.sharedService.loadDataset(
+                tabNumber,
+                plotRef,
+                viewContainerRef,
+                sample
+              );
+            }
+            plotRef.addSamples(samples);
+
```
Comment (agarwal-v) [inline frontend/src/app/side-menu/side-menu.component.ts:94]: unnecessary blank line

File: backend/server.py (hunk 2 above)
Comment (nick-Ag) [reply backend/server.py:85]: The running average looks at the previous k data points to determine the average. How should I determine k? Could I just take some % of the data for k?

Comment (agarwal-v) [reply backend/server.py:85]: Rather than a % you could use 101 points. 50 on either side of the point in question.

Comment (agarwal-v) [reply backend/server.py:85]: Rather than a %, just use last 100 points

Comment (nick-Ag) [reply backend/server.py:85]: But then if the data trace is < 100 points, the running average would be the same as the cumulative avg right?

File: backend/server.py (hunk 1 above)
Comment (rglasser) [reply backend/server.py:65]: Python doc strings should be after the function def, before the code block

def compute_stats():
  """Handles requests for stats data and responds with the requested
      stats.
  Attributes:
      request.data: The data sent from the frontend.
          request.data.channels: Key value pairs of channels and arrays
          to compute statistics for.
  """

**Pull request merged**